src/shims/fs.rs

```diff
@@ -257,6 +264,197 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         this.try_unwrap_io_result(result)
     }
 
+    fn statx(
+        &mut self,
+        dirfd_op: OpTy<'tcx, Tag>, // Has type `int`
+        pathname_op: OpTy<'tcx, Tag>, // Has type `const char *`
+        flags_op: OpTy<'tcx, Tag>, // Has type `int`
+        _mask_op: OpTy<'tcx, Tag>, // Has type `unsigned int`
+        statxbuf_op: OpTy<'tcx, Tag> // Has type `struct statx *`
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        this.check_no_isolation("statx")?;
+
+        let statxbuf_scalar = this.read_scalar(statxbuf_op)?.not_undef()?;
+        let pathname_scalar = this.read_scalar(pathname_op)?.not_undef()?;
+
+        // If the statxbuf or pathname pointers are null, the function fails with `EFAULT`.
+        if this.is_null(statxbuf_scalar)? || this.is_null(pathname_scalar)? {
+            let efault = this.eval_libc("EFAULT")?;
+            this.set_last_error(efault)?;
+            return Ok(-1);
+        }
+
+        // If statxbuf is not null, we need to produce a proper `MemPlace` to write the results of
+        // the function to it. This is needed because all the `statx` parameters are provided as
+        // `isize`s instead of having the proper types. In particular `statxbuf` should be a `*mut
+        // statx`. If this wasn't the case, we could use `deref_operand` instead.
+
+        // This long path is required instead of `libc::statx` because the `statx` struct and
+        // function share the same namespace.
+        let statx_ty = this
+            .resolve_path(&["libc", "unix", "linux_like", "linux", "gnu", "statx"])?
+            .ty(*this.tcx);
+        let statxbuf_ty = this.tcx.mk_mut_ptr(statx_ty);
+        let statxbuf_layout = this.layout_of(statxbuf_ty)?;
+        let statxbuf_imm = ImmTy::from_scalar(statxbuf_scalar, statxbuf_layout);
+        let statxbuf_place = this.ref_to_mplace(statxbuf_imm)?;
+
+        let pathname: PathBuf = this.read_os_str_from_c_str(pathname_scalar)?.into();
+        let flags = this.read_scalar(flags_op)?.to_machine_isize(&*this.tcx)? as i32;
+
+        let at_empty_path = this.eval_libc_i32("AT_EMPTY_PATH")?;
+        let at_fdcwd = this.eval_libc_i32("AT_FDCWD")?;
+
+        // Now we need to resolve the path of the target file. This was done following the manual
+        // entry for `statx`
+
+        // If `pathname` is absolute we can ignore `dirfd`.
+        let path = if pathname.is_absolute() {
+            pathname
+        } else {
+            let dirfd = this.read_scalar(dirfd_op)?.to_machine_isize(&*this.tcx)? as i32;
+            // If `dirfd` is `AT_FDCWD`, `pathname` is relative to the current directory.
+            if dirfd == at_fdcwd {
+                match std::env::current_dir() {
+                    Ok(cwd) => cwd.join(pathname),
```

Comment:
Why is this even needed? If we just use a relative `pathname` ourselves, it *will* be interpreted relative to the current directory, right?

Reply:
The thing is that I'm not sure, the `libstd` code is a bit too exotic for me. I'll read it later and tell you what it does.

Reply:
I mean what else should it do? It seems like a serious bug to me for a relative path to be interpreted in any other way.

Reply:
The manual says that if the file descriptor points to a directory, then the path is relative to that directory, why? idk. The only thing that I can think of is that maybe `statx` was introduced to behave both as `stat` and  `fstat`. 

I remember that there is a flag that `libstd` always uses: `AT_STATX_SYNC_AS_STAT`. The manual says:
```quote 
Do whatever stat(2) does.  This is the default and is very much filesystem-specific.
```
I haven't been able to figure out what does that means. I would be awesome if that meant that `statx` uses the path just like `stat` instead of using the file descriptor as `fstat`.

Reply:
You still didn't explain why you look up `env:current_dir()` here. What's wrong with just calling `fs::metadata` with a relative path?

Reply:
Nothing really. I'll change it